### PR TITLE
Transformer multi-head attention

### PR DESCRIPTION
**Pull request by yushiyangk**
Added modules for scaled dot-product attention and multi-head attention, to be used as components of a Transformer.

Added a method, `T()`, to Tensor to perform simple transposition of the last two axes only of a Tensor. This operation is used in several places in a Transformer, as well as in the updated gradient function for `matmul` (the previous version of this gradient function only supported 2D Tensors).

### CODE REVIEW

**Comment by yushiyangk** [comment]
`traverse_gradients` seems to be trying to add two tensors with different shapes

**Comment by cryptodeal** [comment]
@bwasti, output from from `test-arrayfire-cpu-linux` seems to confirm that's the case.

```sh
native code error: doBinaryOpOrBroadcast: cannot perform operation or broadcasting with tensors of shapes (2, 2) and (2, 4) - dimension mismatch.
                  at _add (/home/circleci/project/shumai/cpp/binding_gen.inl:565)
1276 |         recorded_stat = [performance.now(), fl.bytesUsed()]
1277 |       }
1278 | 
1279 |       const _ptr = fl._add(this.ptr, tensor.ptr)
1280 |       if (!_ptr)
1281 |         throw new Error('Tensor returned from `add` is null; native code likely threw an error...')
                  ^
error: Tensor returned from `add` is null; native code likely threw an error...
      at add (/home/circleci/project/shumai/tensor/tensor_ops_shim_gen.ts:1281:14)
      at traverse_gradients (/home/circleci/project/shumai/tensor/tensor.ts:118:38)
      at backward (/home/circleci/project/shumai/tensor/tensor.ts:238:20)
      at /home/circleci/project/test/transformer.test.ts:159:4
✗ TransformerDotProductAttention > calculates gradient
```

File: shumai/module/transformer.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+export class TransformerDotProductAttention extends Module {
+  private dim: number
+  private scale_factor: Tensor
+
+  constructor(dim: number) {
+    super()
+    this.dim = dim
+    this.scale_factor = sm.scalar(1 / Math.sqrt(dim))
+    this.scale_factor.detach()
```
Comment (bwasti) [inline shumai/module/transformer.ts:18]: this shouldn't be necessary - `requires_grad` defaults to false

File: shumai/module/transformer.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        `Last axis of input tensors (shape ${shape}) must match module dimensions (${this.dim})`
+      )
+    }
+
+    const reshape = [...shape] // [..., tokens, dim]
+    reshape[reshape.length - 1] = this.heads
+    reshape.push(this.attn_dim) // [..., tokens, heads, attn_dim]
+
+    // Swap 2nd and 3rd last axes
+    const transpose = reshape.map((x, i) => i)
```
Comment (bwasti) [inline shumai/module/transformer.ts:106]: not sure its too much cleaner, but you can do `Array.from(sm.util.range(reshape.length))`

**Review by bwasti — APPROVED** [review]
phenomenal work!

**Comment by jacobkahn** [comment]
@yushiyangk this is awesome! Thanks for the hard work!
